File: 19th/코딩테스트 심화반/week8/2211번_네트워크 복구.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

vector<pair<int, int>> adj[1001];
int cost[1001];
int parent[1001]; // 역추적

int N, M;

void dijkstra(int n) {
    fill(cost, cost + N + 1, 123456789); // 모든 노드까지의 거리를 inf로 초기화
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;

    pq.push({ 0, n }); // 이동한 비용의 최솟값, 노드 번호
    cost[n] = 0;

    while (!pq.empty()) {
        int c = pq.top().first; // 이동한 비용의 최솟값
        int x = pq.top().second; // 노드 번호
        pq.pop();

        if (cost[x] < c) continue; // 현재 노드까지 이동한 비용의 최솟값보다 클 경우 continue

        for (int i = 0; i < adj[x].size(); i++) {
            int nx = adj[x][i].first;
            int nc = c + adj[x][i].second; // 현재까지의 이동한 비용 + 다음 노드로 이동할 비용

            // 이동한 비용의 최솟값을 갱신하는 경우
            if (cost[nx] > nc) {
                pq.push({ nc, nx });
                cost[nx] = nc;
                parent[nx] = x; // 어느 경로에서 왔는지 저징
            }
        }
    }
}

int main() {
    // 빠른 입출력
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    cin >> N >> M;

    int A, B, C;
    for (int i = 0; i < M; i++) {
        cin >> A >> B >> C;

        adj[A].push_back({ B, C });
        adj[B].push_back({ A, C });
    }

    dijkstra(1);

    // 최소 개수의 회선만을 복구해야 하기에 간선이 N - 1개인 사이클이 없는 그래프
    cout << N - 1 << "\n";
    for (int i = 2; i <= N; i++) {
        cout << i << " " << parent[i] << "\n"; // 각 노드가 어느 경로에서 왔는지 출력
    }
}

File: 19th/코딩테스트 심화반/week8/2211번_네트워크 복구.py
import heapq

n, m = map(int, input().split())

# 인접 리스트 (양방향 그래프)
adj = [[] for _ in range(n + 1)]
for _ in range(m):
    a, b, c = map(int, input().split())
    adj[a].append((b, c))
    adj[b].append((a, c))

INF = int(1e9)
parent = [0] * (n + 1)  # 각 노드의 최단 경로 트리 상 부모 노드 저장

def dijkstra(start):
    global parent
    distance = [INF] * (n + 1)

    q = []
    heapq.heappush(q, (0, start))  # (거리, 시작 노드)
    distance[start] = 0

    while q:
        dist, now = heapq.heappop(q)
        if distance[now] < dist:
            continue
        for i in adj[now]:
            cost = dist + i[1]
            if cost < distance[i[0]]:
                distance[i[0]] = cost
                parent[i[0]] = now  # 최단 경로 트리 부모 갱신
                heapq.heappush(q, (cost, i[0]))

dijkstra(1)

# 결과 출력: (n-1)개의 간선으로 최단 경로 트리 구성
print(n-1)
for i in range(2, n + 1):
    print(i, parent[i])  # (자식, 부모)